exp/controllers/awsinstancestate_controller.go

```diff
@@ -0,0 +1,237 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const Ec2InstanceStateLabelKey = "ec2-instance-state"
+
+// AwsInstanceStateReconciler reconciles a AwsInstanceState object
+type AwsInstanceStateReconciler struct {
+	client.Client
+	Log               logr.Logger
+	sqsServiceFactory func() sqsiface.SQSAPI
+	queueURLs         sync.Map
+	Endpoints         []scope.ServiceEndpoint
+}
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=awsclusters,verbs=get;list;watch
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=awsmachines,verbs=get;list;watch
+
+func (r *AwsInstanceStateReconciler) getSQSService(region string) (sqsiface.SQSAPI, error) {
+	if r.sqsServiceFactory != nil {
+		return r.sqsServiceFactory(), nil
+	}
+
+	globalScope, err := scope.NewGlobalScope(scope.GlobalScopeParams{
+		ControllerName: "awsinstancestate",
+		Region:         region,
+		Endpoints:      r.Endpoints,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	return scope.NewGlobalSQSClient(globalScope, globalScope), nil
+}
+
+func (r *AwsInstanceStateReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.TODO()
+	_ = r.Log.WithValues("namespace", req.NamespacedName, "awsInstanceState", req.Name)
+
+	// Fetch the AWSCluster instance
+	awsCluster := &infrav1.AWSCluster{}
+	err := r.Get(ctx, req.NamespacedName, awsCluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("cluster not found, removing queue URL", "cluster", req.Name)
+			r.queueURLs.Delete(req.Name)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Handle deleted clusters
+	if !awsCluster.DeletionTimestamp.IsZero() {
+		r.Log.Info("cluster deleted, removing queue URL", "cluster", awsCluster.Name)
+		r.queueURLs.Delete(req.Name)
+	}
+
+	// retrieve queue URL if it isn't already tracked
+	if _, ok := r.queueURLs.Load(awsCluster.Name); !ok {
+		URL, err := r.getQueueURL(awsCluster)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		r.queueURLs.Store(awsCluster.Name, queueParams{region: awsCluster.Spec.Region, URL: URL})
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *AwsInstanceStateReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	go func() {
+		r.watchQueuesForInstanceEvents()
+	}()
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&infrav1.AWSCluster{}).
+		WithOptions(options).
+		WithEventFilter(controllers.PausedPredicates(r.Log)).
+		Complete(r)
+}
+
+func (r *AwsInstanceStateReconciler) watchQueuesForInstanceEvents() {
+	ctx := context.TODO()
+	awsClusterList := &infrav1.AWSClusterList{}
+	if err := r.Client.List(ctx, awsClusterList); err == nil {
+		for i, cluster := range awsClusterList.Items {
+			if URL, err := r.getQueueURL(&awsClusterList.Items[i]); err == nil {
+				r.queueURLs.Store(cluster.Name, queueParams{region: cluster.Spec.Region, URL: URL})
+			}
+		}
+	}
+	for range time.Tick(1 * time.Second) {
```

Comment:
We may hit the api call limits if it is too frequent (assuming it is using the api request tokens too?). Maybe make this a parameter or increase the interval?